crypto/engine/eng_devcrypto.c

```diff
@@ -207,6 +247,22 @@ static int cipher_do_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
     return 1;
 }
 
+static int cipher_ctrl(EVP_CIPHER_CTX *ctx, int type, int p1, void* p2)
+{
+    EVP_CIPHER_CTX *to_ctx = (EVP_CIPHER_CTX *)p2;
+    struct cipher_ctx *cipher_ctx;
+
+    if (type == EVP_CTRL_COPY) {
+        /* when copying the context, a new session needs to be initialized */
+        cipher_ctx = (struct cipher_ctx *)EVP_CIPHER_CTX_get_cipher_data(ctx);
+        return (cipher_ctx == NULL)
+            || cipher_init(to_ctx, cipher_ctx->sess.key, EVP_CIPHER_CTX_iv(ctx),
+                           (cipher_ctx->op == COP_ENCRYPT));
+    }
+
+    return 1;
```

Comment:
If I am not mistaken, other `EVP_*_ctrl()` functions return `-2` when called with an unsupported type.

@levitte definitely knows more, and checking at the documentation is not immediate to understand what should be the return value for this kind of `*_ctrl` functions.

Reply:
You are entirely correct! I totally spaced on this...

Reply:
I was in doubt as well, but looking at `crypto/evp/e_*.c`, the default seems to be -1.  I'll change to that.  If I'm mistaken, let me know.

Reply:
True. Most of all, one can observe that `EVP_CIPHER_CTX_ctrl` checks for -1 specifically.

Other EVP_ APIs use -2 for the same thing, go figure...